lib/action_client/middleware/parser.rb

```diff
@@ -0,0 +1,74 @@
+module ActionClient
+  module Middleware
+    class Parser
+      JsonParser = proc do |body|
+        JSON.parse(body, object_class: HashWithIndifferentAccess)
+      end
+      XmlParser = proc do |body|
+        Nokogiri::XML.parse(body).tap do |document|
+          document.validate
+
+          document.errors.each do |error|
+            if error.is_a?(Nokogiri::XML::SyntaxError)
+              raise error
+            end
+          end
+        end
+      end
+
+      class_attribute :parsers, default: {
+        "application/json": JsonParser,
+        "application/xml": XmlParser,
+      }.with_indifferent_access
+
+      def initialize(app, configuration = {})
+        @app = app
+        @parsers = self.class.parsers.merge(
+          configuration.fetch(:parsers, {}).transform_keys { |content_type| Mime::Type.lookup(content_type).to_s }
+        )
+      end
+
+      def call(env)
+        status, headers, body_proxy = @app.call(env)
+        body = body_proxy.each(&:yield_self).join
+        content_type = headers[Rack::CONTENT_TYPE].to_s
+
+        if body.present?
+          parser = fetch_parser_for_content_type(content_type)
+
+          begin
+            [ status, headers, parser.call(body) ]
+          rescue StandardError => error
+            raise ActionClient::ParseError.new(error, body, content_type)
+          end
+        else
+          [ status, headers, body ]
+        end
+      end
+
+      private
+
+      def fetch_parser_for_content_type(content_type)
+        mime_type = Mime::Type.lookup(content_type)
+
+        @parsers.fetch(mime_type.to_s)
+      rescue Mime::Type::InvalidMimeType, KeyError
```

Comment:
@georgebrock it turns out that `Mime::Type.lookup` will raise if a value does not exist.

Similarly, the `KeyError` will be raised by line 54. We then change the input and recurse in the `rescue` block.

Is this too complicated? Do you have any ideas on how to improve this?